# Title: Customer Churn in Bank Data Analysis

# Problem Statement:

The credit card business of a North American bank is facing declining performance, with a significant number of customers discontinuing their credit card services. To address this issue and reduce customer attrition, the bank needs to gain a deeper understanding of the factors contributing to customer churn.

# Business Objective:

The primary objective is to perform a comprehensive customer churn analysis using available data. This analysis aims to identify patterns, trends, and key drivers of customer attrition, enabling the bank to implement targeted strategies and measures to retain existing customers and improve overall business performance.

# Context and Need for Study:

In today's highly competitive financial industry, retaining customers is essential for sustainable growth. Customer churn can lead to revenue loss and decreased market share. Understanding the underlying reasons for attrition is crucial for developing proactive retention strategies. This study will provide valuable insights into customer behavior, allowing the bank to make data-driven decisions and enhance customer satisfaction and loyalty.

# Understanding the Dataset 

In [2]:
import pandas as pd
Customer_data = pd.read_csv("DS1_C8_V3_ND_Sprint3_BankChurn_Data.csv")
Customer_data


CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0      768805383  Existing Customer            45      M                3   
1      818770008  Existing Customer            49      F                5   
2      713982108  Existing Customer            51      M                3   
3      769911858  Existing Customer            40      F                4   
4      709106358  Existing Customer            40      M                3   
...          ...                ...           ...    ...              ...   
10122  772366833  Existing Customer            50      M                2   
10123  710638233  Attrited Customer            41      M                2   
10124  716506083  Attrited Customer            44      F                1   
10125  717406983  Attrited Customer            30      M                2   
10126  714337233  Attrited Customer            43      F                2   

      Education_Level Marital_Status Income_Category Card_Category  \
0         High School        Married     $60K - $80K          Blue   
1            Graduate         Single  Less than $40K          Blue   
2            Graduate        Married    $80K - $120K          Blue   
3         High School        Unknown  Less than $40K          Blue   
4          Uneducated        Married     $60K - $80K          Blue   
...               ...            ...             ...           ...   
10122        Graduate         Single     $40K - $60K          Blue   
10123         Unknown       Divorced     $40K - $60K          Blue   
10124     High School        Married  Less than $40K          Blue   
10125        Graduate        Unknown     $40K - $60K          Blue   
10126        Graduate        Married  Less than $40K        Silver   

                 Region  Months_on_book  Total_Relationship_Count  \
0               England              39                         5   
1      Northern Ireland              44                         6   
2               England              36                         4   
3                 Wales              34                         3   
4               England              21                         5   
...                 ...             ...                       ...   
10122           England              40                         3   
10123           England              25                         4   
10124           England              36                         5   
10125           England              36                         4   
10126          Scotland              25                         6   

       Months_Inactive_12_mon  Contacts_Count_12_mon  Credit_Limit  \
0                           1                      3       12691.0   
1                           1                      2        8256.0   
2                           1                      0        3418.0   
3                           4                      1        3313.0   
4                           1                      0        4716.0   
...                       ...                    ...           ...   
10122                       2                      3        4003.0   
10123                       2                      3        4277.0   
10124                       3                      4        5409.0   
10125                       3                      3        5281.0   
10126                       2                      4       10388.0   

       Total_Revolving_Bal  Avg_Open_To_Buy  Total_Trans_Amt  Total_Trans_Ct  \
0                      777          11914.0             1144              42   
1                      864           7392.0             1291              33   
2                        0           3418.0             1887              20   
3                     2517            796.0             1171              20   
4                        0           4716.0              816              28   
...                    ...              ...              ...             ...   
10122      

The data is related to customer churn in bank data analysis which consists of 10127 rows and 20 columns

# Preprocessing using Python:

## 1)Display the summary statistics of the dataset.

In [2]:
# Display summary statistics
summary_stats = Customer_data.describe()
print(summary_stats)


          CLIENTNUM  Customer_Age  Dependent_count  Months_on_book  \
count  1.012700e+04  10127.000000     10127.000000    10127.000000   
mean   7.391776e+08     46.382245         2.787400       35.928409   
std    3.690378e+07      8.341393         6.479135        7.986416   
min    7.080821e+08     18.000000         0.000000       13.000000   
25%    7.130368e+08     41.000000         1.000000       31.000000   
50%    7.179264e+08     46.000000         2.000000       36.000000   
75%    7.731435e+08     52.000000         3.000000       40.000000   
max    8.283431e+08     89.000000       143.000000       56.000000   

       Total_Relationship_Count  Months_Inactive_12_mon  \
count              10127.000000            10127.000000   
mean                   3.812580                2.341167   
std                    1.554408                1.010622   
min                    1.000000                0.000000   
25%                    3.000000                2.000000   
50%            

 CLIENTNUM: The unique identifier for customers holding accounts ranges from approximately 708 million to 828 million.

Customer_Age: The average age of customers is approximately 46.4 years, with a minimum age of 18 and a maximum age of 89.

Dependent_count: On average, customers have approximately 2.8 dependents, with some customers having as many as 143 dependents.

Months_on_book: The average duration since customers purchased a loan or credit card is approximately 36 months, with a minimum of 13 months and a maximum of 56 months.

Total_Relationship_Count: On average, customers have around 3.8 relationships with the bank, with a minimum of 1 and a maximum of 6 relationships.

Months_Inactive_12_mon: Customers were inactive for an average of approximately 2.3 months in the last 12 months, with a minimum of 0 months (indicating some customers were active throughout) and a maximum of 6 months.

Contacts_Count_12_mon: On average, customers were contacted by the bank approximately 2.5 times in the last 12 months, with a minimum of 0 contacts and a maximum of 6 contacts.

Credit_Limit: The average credit limit offered by the bank is around $8703, with a minimum of $1438.30 and a maximum of $34664.

Total_Revolving_Bal: On average, customers carry a revolving balance of approximately 1162.81, with some customers having no revolving balance and others having up to 2517 in revolving balance.

Avg_Open_To_Buy: The average available credit for customers is approximately $7469.14, with a minimum of $3 and a maximum of $34516.

Total_Trans_Amt: Customers conducted an average of $4404.09 in total transaction amounts, with a minimum of $510 and a maximum of $18484.

Total_Trans_Ct: The average total number of transactions by customers in the last 12 months is approximately 64.86, with a minimum of 10 transactions and a maximum of 139 transactions.

Avg_Utilization_Ratio: The average utilization ratio of customers' credit cards is about 27.49%, with values ranging from 0% to 99.9%.

These statistics provide an overview of the customer data, including age distribution, relationship counts, credit limits, transaction amounts, and more, which can be further analyzed to gain insights into customer behavior and attrition.

## 2)Identify the outliers in the dataset.



In [3]:
def detect_outliers_iqr_all(data):
    numeric_columns = data.select_dtypes(include=['number']).columns
    all_outliers = pd.DataFrame()

    for column in numeric_columns:
        Q1 = data[column].quantile(0.25)
        Q3 = data[column].quantile(0.75)
        IQR = Q3 - Q1
        lower_bound = Q1 - 1.5 * IQR
        upper_bound = Q3 + 1.5 * IQR
        outliers = data[(data[column] < lower_bound) | (data[column] > upper_bound)]
        outliers['Outlier_Column'] = column
        all_outliers = pd.concat([all_outliers, outliers])

    return all_outliers

# Detect outliers for all numeric columns
all_outliers = detect_outliers_iqr_all(Customer_data)

# Display the outliers for all numeric columns
print("All Outliers:")
print(all_outliers)


All Outliers:
       CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
60     787584108  Existing Customer            77      M                3   
72     719580033  Existing Customer            87      M                1   
198    708868233  Attrited Customer            23      M                0   
251    715952883  Existing Customer            73      M                0   
254    787348608  Existing Customer            70      M                0   
...          ...                ...           ...    ...              ...   
10123  710638233  Attrited Customer            41      M                2   
10124  716506083  Attrited Customer            44      F                1   
10126  714337233  Attrited Customer            43      F                2   
9324   708163758  Existing Customer            41      M                3   
9586   784868958  Existing Customer            56      F                1   

      Education_Level Marital_Status Income_Category Card_Cat

C:\Users\hp\AppData\Local\Temp\ipykernel_10324\3449890452.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  outliers['Outlier_Column'] = column
C:\Users\hp\AppData\Local\Temp\ipykernel_10324\3449890452.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  outliers['Outlier_Column'] = column
C:\Users\hp\AppData\Local\Temp\ipykernel_10324\3449890452.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See 

The function will iterate through all numeric columns, detect outliers for each column, and store them in a DataFrame called all_outliers. Each row in this DataFrame will indicate the outlier values and the column in which the outlier was detected.

## 3) Identify the missing values in the dataset and display the count of missing values.

In [4]:
# Identify missing values and count
missing_values = Customer_data.isnull().sum()
print(missing_values)


CLIENTNUM                     0
Attrition_Flag                0
Customer_Age                  0
Gender                        0
Dependent_count               0
Education_Level               0
Marital_Status                0
Income_Category               0
Card_Category               132
Region                        0
Months_on_book                0
Total_Relationship_Count      0
Months_Inactive_12_mon        0
Contacts_Count_12_mon         0
Credit_Limit                 60
Total_Revolving_Bal           0
Avg_Open_To_Buy               0
Total_Trans_Amt               0
Total_Trans_Ct                0
Avg_Utilization_Ratio         0
dtype: int64


There are no missing values in the dataset except the Card_Category column 

## 4) Perform data imputation, using whichever method suits most logically for the given columns in the dataset.

In [6]:
import pandas as pd

# Load your dataset
Customer_data = pd.read_csv("DS1_C8_V3_ND_Sprint3_BankChurn_Data.csv")

# Function to impute missing values in categorical columns with the mode
def impute_categorical(df):
    categorical_columns = df.select_dtypes(include=['object']).columns
    for col in categorical_columns:
        df[col].fillna(df[col].mode()[0], inplace=True)

# Function to impute missing values in numeric columns with the median
def impute_numeric(df):
    numeric_columns = df.select_dtypes(include=['number']).columns
    for col in numeric_columns:
        df[col].fillna(df[col].median(), inplace=True)

# Call the functions to impute missing values
impute_categorical(Customer_data)
impute_numeric(Customer_data)

missing_values = Customer_data.isnull().sum()
print(missing_values)

CLIENTNUM                   0
Attrition_Flag              0
Customer_Age                0
Gender                      0
Dependent_count             0
Education_Level             0
Marital_Status              0
Income_Category             0
Card_Category               0
Region                      0
Months_on_book              0
Total_Relationship_Count    0
Months_Inactive_12_mon      0
Contacts_Count_12_mon       0
Credit_Limit                0
Total_Revolving_Bal         0
Avg_Open_To_Buy             0
Total_Trans_Amt             0
Total_Trans_Ct              0
Avg_Utilization_Ratio       0
dtype: int64


All missing Values in the dataset are been imputed by the Statistical methods like mean,median and mode

## To use your cleaned data in Tableau, you can export it to a CSV file using the to_csv method provided by pandas

In [7]:
# Export the cleaned data to a CSV file
Customer_data.to_csv("cleaned_customer_data.csv", index=False)


# cleaned_customer_data.csv is the file that can be used for vizualizations in the tableau 